docs: AWS KMS (#37388)
gravitational/teleport: docs/pages/choose-an-edition/teleport-enterprise/aws-kms.mdx

---
title: AWS KMS
description: Configure Teleport to store CA private keys in the AWS Key Management Service
h1: Store Teleport Private Keys in AWS KMS
---

> This guide will show you how to set up your Teleport Cluster to use the AWS Key
> Management Service (KMS) to store and handle the CA private key material used to
> sign all certificates issued by your Teleport cluster.

Reviewer: ```suggestion
This guide will show you how to set up your Teleport cluster to use the AWS Key
Management Service (KMS) to store and handle the CA private key material used to
sign all certificates issued by your cluster.
```
- Capitalization
- Fix repetition

Teleport generates private key material for its internal Certificate Authorities

> (CAs) during the first Auth Server's initial startup.

Reviewer: ```suggestion
(CAs) during the first Auth Service instance's initial startup.
```
For consistency with the rest of the docs

These CAs are used to sign all certificates issued to clients and hosts in the
Teleport cluster.
When configured to use AWS KMS, all private key material for these CAs
will be generated, stored, and used for signing inside of AWS KMS.
Instead of the actual private key, Teleport will only store the ID of the KMS
key.
In short, private key material will never leave AWS KMS.

If launching a new Teleport cluster this will all be handled during initial
startup with no specific interventions required after configuration.

> For existing Teleport clusters that already have private key material on disk,

Reviewer 2: I suspect most people interested in this feature will be those using Teleport on AWS with DynamoDB, so the "already on disk" point isn't entirely accurate..

            ],
            "Resource": "*",
            "Condition": {
                "StringEquals": {
                    "aws:RequestTag/TeleportCluster": "<Var name="teleport.example.com" description="Your Teleport cluster name"/>",
                    "aws:ResourceTag/TeleportCluster": "<Var name="teleport.example.com" description="Your Teleport cluster name"/>"
                }
            }
        },
        {
            "Sid": "UseKeys",
            "Effect": "Allow",
            "Action": [
                "kms:GetPublicKey",
                "kms:ScheduleKeyDeletion",
                "kms:DescribeKey",
                "kms:ListResourceTags",
                "kms:Sign"
            ],
            "Resource": "*",
            "Condition": {
                "StringEquals": {
                    "aws:ResourceTag/TeleportCluster": "<Var name="teleport.example.com" description="Your Teleport cluster name"/>"
                }
            }
        }
    ]
}
```

You can create the policy from the AWS console, or via the AWS CLI.

```code
$ aws iam create-policy --policy-name TeleportKMS --policy-document file://policy.json
```

### Attach the AWS IAM policy

> Your Teleport Auth Server will need access to an IAM identity with this policy attached.

Reviewer: ```suggestion
Your Teleport Auth Service will need access to an IAM identity with this policy attached.
```

You can provide credentials via IAM Roles for Tasks, IAM Roles for EC2,
credential files, or environment variables.

> See [Specifying Credentials](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specifying-credentials)

Reviewer: Would the `docs/pages/includes/aws-credentials.mdx` partial provide this information?

Author: yeah, didn't know about that, thanks

for details.

If your Teleport Auth Service is running on an EC2 instance and already has an

> attached "IAM role for Amazon EC2", attach the above policy to the existing role.

Reviewer: What do the quotes mean here?

Author: this got replaced by the partial

If the instance does not already have an attached role, create an IAM role with
the above policy and attach it to your EC2 instance running the Teleport Auth
Service.

> ## Step 2/5. Configure your Auth Server to use KMS keys

Reviewer: ```suggestion
## Step 2/5. Configure your Auth Service to use KMS keys
```

CA key parameters are statically configured in the `teleport.yaml` configuration

> file of the Teleport Auth Services in your cluster.

Reviewer: ```suggestion
file of the Teleport Auth Service instances in your cluster.
```

$ tctl get cert_authorities --with-secrets --format json | jq -r '.[].spec.active_keys.jwt[0].private_key | select(.) | @base64d'
awskms:arn:aws:kms:us-west-2:123456789012:key/d4e63a5d-ed81-46c4-ad72-3f57a07eb9d5
awskms:arn:aws:kms:us-west-2:123456789012:key/5c99e6aa-deb1-45c0-865e-08847214b7ab
```

Try logging in to the cluster with a Teleport user to make sure that new
certificates can be signed without error.

## Migrating an existing cluster

If you have an existing Teleport cluster it will have already created software
CA keys during its first start.
Those existing CA keys will have been used to sign all existing user and host
certificates, and will be trusted by all other services in your cluster.

In order for the Teleport Auth Service to generate new keys in AWS KMS and have
them trusted by the rest of the cluster, you will need to rotate all of your
Teleport CAs.

`teleport start` will print a warning during startup if any CA needs to be rotated.
Any users allowed the `update` verb for `cert_authority` resources in any of
their Teleport roles will also see a cluster alert reminding them to rotate the
CAs.

> CA rotation can be performed manually or semi-automatically, see our admin guide

Reviewer: ```suggestion
CA rotation can be performed manually or semi-automatically. See our admin guide
```
